Add DeiT Model

Sohaib-Ahmed21 opened the pull request:
Add VIT based DeiT (data-efficient image transformers) model to keras-hub along with its backbone, layers, tests. 
Paper: [https://arxiv.org/pdf/2012.12877](https://arxiv.org/pdf/2012.12877)
Model card: [https://huggingface.co/facebook/deit-small-distilled-patch16-224](https://huggingface.co/facebook/deit-small-distilled-patch16-224)

Sohaib-Ahmed21 commented:
Kindly approve the workflows.

sachinprasadhs commented:
@Sohaib-Ahmed21 , Thanks for the PR, could you please add the `checkpoint_conversion` , and link a colab notebook here verifying the numerics, parameters matching and demonstrating end to end usage example of the model.

Sohaib-Ahmed21 commented:
> @Sohaib-Ahmed21 , Thanks for the PR, could you please add the `checkpoint_conversion` , and link a colab notebook here verifying the numerics, parameters matching and demonstrating end to end usage example of the model.

Yeah sure, I'll do that and update the PR, thanks!

Sohaib-Ahmed21 commented:
Added the `checkpoint_conversion`, will share the parameter/numerics verification and end to end demo notebooks soon.

Sohaib-Ahmed21 commented:
I’m sharing the following resources related to this PR:

- DeiT Checkpoint Conversion and Numerics Verification Demo (across multiple backends): [Notebook Link](https://colab.research.google.com/drive/15DIJm-pha9LUSOHMZaYB_Ardl4vQXPmK#scrollTo=JU_C_IBGL_WC)
- DeiT End-to-End Demo (zero-shot and finetuning): [Notebook Link](https://colab.research.google.com/drive/1eGSRfXrfKknUhp0n1nQ62mIQKhjtY3pq#scrollTo=Wjtz0e6j0mr-)
- [Here](https://drive.google.com/drive/folders/1ni6io9mMzm0lZLv0jwFUv0Q6x2DZ3MzK?usp=drive_link) are the converted DeiT presets from Hugging Face checkpoints for reference.

Sohaib-Ahmed21 commented:
This PR is ready for review. Kindly approve the workflows and review the PR, thanks!

sachinprasadhs commented on keras_hub/src/models/deit/deit_backbone.py:
```diff
@@ -0,0 +1,158 @@
+
+    Args:
+        image_shape: A tuple or list of 3 integers representing the shape of the
+            input image `(height, width, channels)`, `height` and `width` must
```
Can we make this flexible to handle different height and width inputs.
For the weights you are transferring you can anyways provide same height and width.
For other cases or models if users wants to use this as Encoder, it should be flexible enough to handle different height and width.

sachinprasadhs reviewed (commented):
Thanks for the demo notebook, added few more comments.

Also, in the colab can you assert the parameters matching between Keras and HF model.
Example, tiny variant has "params": 5524800, can you show the output where it matches with the HF params?

mattdangerw commented on keras_hub/src/layers/preprocessing/image_converter.py:
```diff
@@ -197,7 +197,7 @@ def __init__(
         scale=None,
         offset=None,
-        crop_to_aspect_ratio=True,
+        crop_to_aspect_ratio=False,
```
We probably want to keep this true for all models, including Deit!

crop_to_aspect_ratio is the default behavior of all our image backbones that can be used for classification. it's an arbitrary choice of a good default really, you could override this from user side code for any model, and use a `CenterCrop` layer from Keras in the preprocessing pipeline. We shouldn't need special logic for that here.

Sohaib-Ahmed21 replied:
I'll revert it for backward compatibility. I changed it because default `True` didn't match the outputs of keras and hf preprocessors and the comment [here](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/layers/preprocessing/image_converter.py#L155C1-L156C1) says default is False. I'll pass `False` in my call and adjust [this](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/layers/preprocessing/image_converter.py#L155C1-L156C1) comment also.

mattdangerw commented on keras_hub/src/models/deit/deit_image_converter.py:
```diff
@@ -0,0 +1,211 @@
+
+
+@keras_hub_export("keras_hub.layers.DeiTImageConverter")
+class DeiTImageConverter(ImageConverter):
```
we likely want this be an empty subclass like other classifier backbones. the base class can handle scale and offset (yes we have to convert mean and std params from other libraries to these), and basic resizing. anything else can be done in more custom preprocessing pipelines. Might be missing something but i don't think there is anything to Deit and center cropping images (vs resizing them).

Sohaib-Ahmed21 replied:
Ok, I'll make subclass empty as base class seems logically sufficient.

Sohaib-Ahmed21 replied:
Made subclass empty and updated `preprocessor conversion` from HF checkpoint accordingly.

Sohaib-Ahmed21 commented:
I've addressed all reviews, kindly review the updated PR. The [notebook](https://colab.research.google.com/drive/15DIJm-pha9LUSOHMZaYB_Ardl4vQXPmK#scrollTo=KWzfPF21RNHi) has also been updated to include parameter verification.

Sohaib-Ahmed21 commented:
Is the failing test related to the PR? Kindly confirm and re-run the tests if required.

mattdangerw commented:
@Sohaib-Ahmed21 no the jax-gpu segfault popped up recently, it's probably related to our test environment (we haven't tracked it down yet). You can ignore.

mattdangerw commented on keras_hub/src/models/deit/deit_presets.py:
```diff
@@ -0,0 +1,49 @@
+
+# Metadata for loading pretrained model weights.
+backbone_presets = {
+    "deit-base-distilled-patch16-384_imagenet": {
```
I think we use only underscore in our preset names, no mixing dashes and underscore like this.

mattdangerw replied:
This should be true for the preset names here and on Kaggle, we want consistency.

Sohaib-Ahmed21 replied:
Used underscore in preset names for consistency with other models.

mattdangerw commented on keras_hub/src/utils/transformers/preset_loader.py:
```diff
@@ -73,7 +76,10 @@ def load_task(self, cls, load_weights, load_task_weights, **kwargs):
             )
         # Support loading the classification head for classifier models.
-        if architecture == "ViTForImageClassification":
+        if (
```
Maybe we could just check if `ForImageClassification` is in the name here?

mattdangerw reviewed (commented):
Few misc comments, but this is looking good!

Sohaib-Ahmed21 commented:
Failing test seems unrelated to PR, kindly review, thanks!

JyotinderSingh commented:
> I’m sharing the following resources related to this PR:
> 
> * DeiT Checkpoint Conversion and Numerics Verification Demo (across multiple backends): [Notebook Link](https://colab.research.google.com/drive/15DIJm-pha9LUSOHMZaYB_Ardl4vQXPmK#scrollTo=JU_C_IBGL_WC)
> * DeiT End-to-End Demo (zero-shot and finetuning): [Notebook Link](https://colab.research.google.com/drive/1eGSRfXrfKknUhp0n1nQ62mIQKhjtY3pq#scrollTo=Wjtz0e6j0mr-)
> * [Here](https://drive.google.com/drive/folders/1ni6io9mMzm0lZLv0jwFUv0Q6x2DZ3MzK?usp=drive_link) are the converted DeiT presets from Hugging Face checkpoints for reference.

Please add these links to the PR description to make them easier to find.

Sohaib-Ahmed21 commented:
This PR seems ready for merge. Kindly review/merge it. Thanks!

divyashreepathihalli commented:
In the demo colab
- why are you using the converter outside of the task class?
- this should be included in the task preprocessor right?

sachinprasadhs commented on tools/checkpoint_conversion/convert_deit_checkpoints.py:
```diff
@@ -0,0 +1,404 @@
+    return DeiTImageConverter(
+        image_size=image_size,
+        scale=config["rescale_factor"],
+        mean=mean,
+        variance=[s**2 for s in std],  # variance
```
No need to pass this, 
just pass `offset` and `scale`
offset is calculated using:
`offset = [-m / s for m, s in zip(mean, std)]`

Sohaib-Ahmed21 replied:
[Here](https://github.com/keras-team/keras-hub/blob/4bed9ae27db0e0b1b8621c31a320a1c1c08b047c/keras_hub/src/layers/preprocessing/image_converter.py#L212), std is calculated (from variance passed above) which is used in [offset](https://github.com/keras-team/keras-hub/blob/4bed9ae27db0e0b1b8621c31a320a1c1c08b047c/keras_hub/src/layers/preprocessing/image_converter.py#L214) calculated using above formula.

sachinprasadhs replied:
You need to pass calculated `scale` and `offset` to the `DeiTImageConverter` which are part of the argument here https://github.com/keras-team/keras-hub/blob/4bed9ae27db0e0b1b8621c31a320a1c1c08b047c/keras_hub/src/layers/preprocessing/image_converter.py#L198-L199.

The scale value which you are passing is wrong, it should be `scale = [config["rescale_factor"]/ s for s in std]`
And offset is calculated as I mentioned above,  in future we will remove taking "mean" and "variance" as part of arguments.

sachinprasadhs commented:
I was looking into your generated files, especially in `preprocessor.json`, how are you populating `norm_mean` and `norm_std`

Sohaib-Ahmed21 commented:
> In the demo colab
> 
> * why are you using the converter outside of the task class?
> * this should be included in the task preprocessor right?

Other models also define the converter separately and link it to the preprocessor (e.g., ViT in keras-hub uses converter outside the task class [here](https://github.com/keras-team/keras-hub/blob/4bed9ae27db0e0b1b8621c31a320a1c1c08b047c/keras_hub/src/models/vit/vit_image_converter.py#L36))

Sohaib-Ahmed21 commented:
> I was looking into your generated files, especially in `preprocessor.json`, how are you populating `norm_mean` and `norm_std`

I have updated the presets at the [link](https://drive.google.com/drive/folders/1ni6io9mMzm0lZLv0jwFUv0Q6x2DZ3MzK?usp=drive_link) w.r.t new changes, now `norm_mean` and `norm_std` are not in `preprocessor.json`.

Sohaib-Ahmed21 commented:
This PR seems ready for merge. Kindly review/merge it. Thanks!

Sohaib-Ahmed21 commented:
Kindly review/merge this pull request, thanks!
